bpf: fix flakes when checking metrics map values.

tommyp1ckles opened the pull request:
The ipsec_from_lxc_test and tc_lxc_policy_drop test files both perform assertions on expected count values stored in the metric map. This results in test flakes as if the map put and map get are performed while running on different cpus then the assertions may fail.

This introduces a new helper macro in tests which iterates possible percpu hash values for a particular key until an expected count is found or we exceed the number of cpu slots.

Fixes: #28280

tommyp1ckles committed:
bpf: fix flakes when checking metrics map values.

The ipsec_from_lxc_test and tc_lxc_policy_drop test files both perform
assertions on expected count values stored in the metric map.
This results in test flakes as if the map put and map get are performed
while running on different cpus then the assertions may fail.

This introduces a new helper macro in tests which iterates possible
percpu hash values for a particular key until we exceed the number of cpu slots
and checks the sum against the desired count.

Fixes: #28280

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

jschwinger233 commented on bpf/tests/common.h:
```diff
@@ -272,4 +272,27 @@ _Pragma("unroll")						\
+		if (!__entry) { \
+			break; \
+		} \
+		if (__entry->count == count) { \
```
Maybe we should accumulate `__entry->count` and compare the `count` with summed up `__entry->count`?

Something like:
```suggestion
__u64 sum;
for (int i = 0; i < 128; i++) {
    __entry = map_lookup_percpu_elem(&METRICS_MAP, &key, i);
    if (!__entry) break;
    sum += __entry->count;
}
if (sum == count) found=true;
```

jschwinger233 replied:
I think it depends on the semantics of macro. The macro we create here is `assert_metrics_contains`, so your implementation above is accurate. My suggested code is more like `assert_metrics_count`.

tommyp1ckles replied:
This makes sense, made the change 👍

jschwinger233 reviewed (commented):
Thanks Tom! Just a small question regarding macro semantic.

gentoo-root commented on bpf/tests/common.h:
```diff
@@ -272,4 +272,27 @@ _Pragma("unroll")						\
 	return 0;						\
 }
 
+static void *(*map_lookup_percpu_elem)(void *map, const void *key, unsigned int cpu) = (void *)195;
```
Any reason not to use BPF_FUNC and put it in bpf/include/bpf/helpers.h?

tommyp1ckles replied:
I'm using the cilium-builder docker image to build this, I get:

```
/home/tom/go/src/github.com/cilium/cilium/bpf/include/linux/../bpf/helpers.h:38:13: error: use of undeclared identifier 'BPF_FUNC_map_lookup_percpu_elem'
static void BPF_FUNC(map_lookup_percpu_elem, const void *map, const void *key, unsigned int cpu);
```

It seems like the version of `linux/bpf.h` we're using doesn't include that definition.

tommyp1ckles replied:
@gentoo-root not sure the best course of action here, any advice?

gentoo-root replied:
We can update bpf.h, we do that when needed:

https://github.com/cilium/cilium/commit/049a3a9da86e5221867013ba99a7bbb581b1b131
https://github.com/cilium/cilium/commit/271a96da32113b9c02465aca3e2587584164924a

tommyp1ckles replied:
Added a commit to sync with latest upstream commit, there's a few changes in the diff - I assume those are all ok/compatible?

tommyp1ckles replied:
@gentoo-root

gentoo-root replied:
Yeah, should be good, as long as the tests don't start failing due to this.

gentoo-root commented on bpf/tests/common.h:
```diff
@@ -272,4 +272,25 @@ _Pragma("unroll")						\
+	__u64 sum = 0; \
+	/* Iterate until lookup encounters null when hitting cpu number */ \
+	/* Assumes at most 128 CPUS */ \
+	for (int i = 0; i < 128; i++) { \
```
Let's maybe `#define NR_CPUS 128` somewhere in a header file? If we can't determine the real number of CPUs, it would still be better to define this constant with the standard name (as used in the kernel), than bury it deep in the code.

tommyp1ckles committed:
bpf: Sync linux/bpf.h to upstream

Newer header is required for definition of map_lookup_percpu_elem.

[0] Upstream commit: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/include/uapi/linux/bpf.h?id=9ed22ae6be817d7a3f5c15ca22cbc9d3963b481d

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

jschwinger233 commented on bpf/tests/common.h:
```diff
@@ -272,4 +275,22 @@ _Pragma("unroll")						\
+			break; \
+		} \
+		sum++; \
+		break; \
```
This `break` limits `sum` from increasing beyond 1, is it intentional?

And maybe we should `sum += __entry->count` instead of `sum++`? Otherwise sum is always set to the number of CPUs.

tommyp1ckles replied:
Woops yup thanks 😅 , fixed

jschwinger233 reviewed (approved):
Thanks!
